Regression/SLM/src/main/java/org/tribuo/regression/slm/SparseLinearModel.java

```diff
@@ -189,14 +201,149 @@ protected DimensionTuple scoreDimension(int dimensionIdx, SparseVector features)
+    }
+
+    /**
+     * Builds a TensorProto containing the model biases.
+     * @param context The naming context.
+     * @return The bias TensorProto.
+     */
+    protected OnnxMl.TensorProto biasBuilder(ONNXContext context) {
+        OnnxMl.TensorProto.Builder biasBuilder = OnnxMl.TensorProto.newBuilder();
+        biasBuilder.setName(context.generateUniqueName("slm_biases"));
+        biasBuilder.addDims(outputIDInfo.size());
+        biasBuilder.setDataType(OnnxMl.TensorProto.DataType.FLOAT.getNumber());
+        ByteBuffer buffer = ByteBuffer.allocate(outputIDInfo.size()*4).order(ByteOrder.LITTLE_ENDIAN);
+        FloatBuffer floatBuffer = buffer.asFloatBuffer();
+        for (int i = 0; i < weights.length; i++) {
+            if (bias) {
+                floatBuffer.put((float)weights[i].get(featureIDMap.size()));
+            } else {
+                floatBuffer.put(0.0f);
+            }
+        }
+        floatBuffer.rewind();
+        biasBuilder.setRawData(ByteString.copyFrom(buffer));
+        return biasBuilder.build();
+    }
+
+    @Override
+    public OnnxMl.GraphProto exportONNXGraph(ONNXContext context) {
+        OnnxMl.GraphProto.Builder graphBuilder = OnnxMl.GraphProto.newBuilder();
+
+        // Make inputs and outputs
+        OnnxMl.TypeProto inputType = ONNXUtils.buildTensorTypeNode(new ONNXShape(new long[]{-1, featureIDMap.size()}, new String[]{"batch", null}), OnnxMl.TensorProto.DataType.FLOAT);
+        OnnxMl.ValueInfoProto inputValueProto = OnnxMl.ValueInfoProto.newBuilder().setType(inputType).setName("input").build();
+        graphBuilder.addInput(inputValueProto);
+        OnnxMl.TypeProto outputType = ONNXUtils.buildTensorTypeNode(new ONNXShape(new long[]{-1, outputIDInfo.size()}, new String[]{"batch", null}), OnnxMl.TensorProto.DataType.FLOAT);
+        OnnxMl.ValueInfoProto outputValueProto = OnnxMl.ValueInfoProto.newBuilder().setType(outputType).setName("output").build();
+        graphBuilder.addOutput(outputValueProto);
+
+        // Add weights
+        OnnxMl.TensorProto weightInitializerProto = weightBuilder(context);
+        graphBuilder.addInitializer(weightInitializerProto);
+
+        // Add biases
+        OnnxMl.TensorProto biasInitializerProto = biasBuilder(context);
+        graphBuilder.addInitializer(biasInitializerProto);
+
+        // Add feature and output means
+        double[] xMean = bias ? Arrays.copyOf(featureMeans.toArray(),featureIDMap.size()) : featureMeans.toArray();
+        OnnxMl.TensorProto featureMeanProto = ONNXUtils.arrayBuilder(context, "feature_mean",xMean);
+        graphBuilder.addInitializer(featureMeanProto);
+        OnnxMl.TensorProto outputMeanProto = ONNXUtils.arrayBuilder(context,"y_mean",yMean);
+        graphBuilder.addInitializer(outputMeanProto);
+
+        // Add feature and output variances
+        double[] xVariance = bias ? Arrays.copyOf(featureVariance.toArray(),featureIDMap.size()) : featureVariance.toArray();
```

Comment:
How does ONNX handle the potential `NaN` values here?

Reply:
The variances are initialized to 1 if they are unused, and if you hit a zero variance feature something else is likely to have blown up before you got to the model part. I don't actually know what ONNX would do if there was a NaN here, I'd expect it to propagate it.